Scenes/Player/PlayerState.cs

```diff
@@ -381,4 +380,84 @@ public class GrappleState : PlayerState {
         }
         return new FallState();
     }
+}
+
+public class GrappleState : PlayerState {
+    public override string Name { get; set; } = "Grapple";
+
+    private readonly GrappleHook _grapple;
+
+    private float _curAngle;
+    private float _angleVel;
+    private float _angleAcc;
+
+    private readonly float _length;
+    private readonly float _gravity = ProjectSettings.GetSetting("physics/2d/default_gravity").AsSingle();
+
+    private readonly Vector2 _entryVelocity;
+
+    public GrappleState(Player player) {
+        _entryVelocity = player.Velocity;
+        player.Velocity = Vector2.Zero;
+
+        _grapple = player.GrappleInstance;
+        var playerPos = player.GlobalPosition;
+
+        _curAngle = (float)Math.PI / 2 - _grapple.GlobalPosition.AngleToPoint(playerPos);
+        _length = playerPos.DistanceTo(_grapple.GlobalPosition);
+    }
+
+    public override PlayerState HandleInput(Player player, Player.InputInfo inputs, double delta) {
+        if (_grapple == null || !inputs.IsPushingGrapple) {
+            // Calculate tangential exit velocity
+            var exitVelocityDirection =
+                new Vector2((float)Math.Cos(_curAngle), (float)Math.Sin(_curAngle)).Normalized();
+            var tangentialExitVelocity = exitVelocityDirection * _angleVel * _length;
+
+            if (_angleVel < 0 || (_angleVel > 0 && _angleAcc < 0))
+                tangentialExitVelocity.Y *= -1;
+
+            player.Velocity = tangentialExitVelocity;
+
+
+            if (player.IsOnFloor()) {
+                return inputs.InputDirection.X != 0 ? new RunState() : new IdleState();
+            }
+            return new FallState();
+        }
+
+        if (player.IsOnFloor() || player.IsOnWall() || player.IsOnCeiling()) {
```

Comment:
Can we use `MoveAndCollide` to help handle this more generically (in particular, for colliding with enemies too)? Should check collisions using the normal system but without `Velocity`. Can move this stuff to L460 using that maybe.